poll/poll_linux.go

```diff
@@ -30,22 +55,67 @@ func New(fd int) (p *Poller, err error) {
 }
 
 func (p *Poller) Close() error {
-	return syscall.Close(p.epfd)
-}
+	p.wakeMutex.Lock()
+	err1 := p.wake.Close()
+	// set wake to nil to be sure that we won't call write on closed wake
+	// it should naver happen but if semeone changes something this might show a bug
+	p.wake = nil
+	p.wakeMutex.Unlock()
 
-func (p *Poller) WaitWrite(deadline time.Time) error {
-	msec := -1
-	if !deadline.IsZero() {
-		d := deadline.Sub(time.Now())
-		msec = int(d.Nanoseconds() / 1000000) // ms!? omg...
+	err2 := syscall.Close(p.epfd)
+	if err1 != nil {
+		return err1
+	} else {
+		return err2
 	}
+}
+
+func (p *Poller) WaitWriteCtx(ctx context.Context) error {
+	doneChan := make(chan struct{})
+	defer close(doneChan)
 
-	n, err := syscall.EpollWait(p.epfd, p.events[:], msec)
+	go func() {
+		select {
+		case <-doneChan:
+			return
+		case <-ctx.Done():
+			select {
+			case <-doneChan:
+				// if we re done with this function do not write to p.wake
+				// it might be already closed and the fd could be reopened for
+				// different purpose
+				return
+			default:
+			}
+			p.wakeMutex.Lock()
+			p.wake.WriteEvents(1) // send event to wake up epoll
```

Comment:
youll want to check if wake is nil here. Since it can be closed and nill'ed in `Close`